apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/MetricSet.java

```diff
@@ -46,6 +47,9 @@ public class MetricSet implements Recyclable {
     // low load factor as hash collisions are quite costly when tracking breakdown metrics
     private final ConcurrentMap<String, Timer> timers = new ConcurrentHashMap<>(32, 0.5f, Runtime.getRuntime().availableProcessors());
     private final ConcurrentMap<String, AtomicLong> counters = new ConcurrentHashMap<>(32, 0.5f, Runtime.getRuntime().availableProcessors());
+
+    //no need for concurrency because this field is only updated from the metrics reporter thread.
+    private final Map<String, Double> rawValues = new HashMap<>();
```

Comment:
But concurrency is not only about updates being from one thread, it's about reads and updates. So for example one known concurrency bug with using HashMap when it's updated in only one thread is that the update could rehash the table while the other thread is iterating. Just switch to a ConcurrentHashMap, true it has overheads until about Java 9, but by 11 it's really low overhead (both CPU and memory) compared to HashMap, so I think it's fine to use

Reply:
My comment here was inaccurate, `rawValues` is updated and read from a single thread.
However, I agree that we should change it to `ConcurrentMap` as this assumption might not hold in the future and is not worth the headache.